### ISSUE TITLE
SNV Validation/merging adjacent SNVs to MNV

### ISSUE THREAD

**Issue by gabeng** [bcbio/bcbio-nextgen#2519]
Hi Brad,

I was exloring the new GATK4 option `--max-mnp-distance` the other day. 
https://software.broadinstitute.org/gatk/documentation/tooldocs/4.0.8.1/org_broadinstitute_hellbender_tools_walkers_haplotypecaller_HaplotypeCaller.php#--max-mnp-distance

Setting this to, say, `2` this may have certain advantages when annotating variants with protein changes. When looking at validation results, however, performance seems to drop. In my view this is due to the fact that an MNV which is now composed of several previously independent SNVs may become entirely a false call if any of those individual SNVs was false. This effect is mentioned at the end of the `vcfeval` tool documentation:
https://cdn.rawgit.com/RealTimeGenomics/rtg-tools/master/installer/resources/tools/RTGOperationsManual/rtg_command_reference.html#vcfeval

My question is: Would it make sense to perform an `rtg vcfdecompose --break-mnps` to improve the validation performance? 
I admit that the right place to meddle with this would rather be the protein change annotation that should be based on an entire group of phased SNVs. The GATK4 option just seemed like a quick fix.

Your feedback is appreciated.

**Comment by chapmanb** [comment]
Ben;
Thanks for investigating this. In general we attempt to break apart MNPs into phased single events rather than try to keep them combined. As you're finding with rtg, there are many inconsistencies in how downstream tools handle MNPs, so dealing with the primitives provides a more standard experience.

I'm agreed with your suggestion that you might need to consider phasing in interpreting effects changes, rather than trying to force MNPs through and potentially hitting issues with validating and representing them. I'd be interested to hear what approaches you come up with to help with interpretation. Thanks again for the discussion.

**Comment by zhengyonglan** [comment]
I am thinking about this too these days, thank you so much for the valuable comments from both of you!

**Comment by gabeng** [comment]
@zhengyonglan BTW: after going into more detail in the results I found that the reason for dropping sensitivity is that GATK is creating overlapping genotypes with these modified settings. You may consider this a "bug" or limitation in `rtg vcfeval`. It cannot handle overlapping genotypes.
I am looking for (clinical) examples that could shed some more light on which representation makes more sense.

**Comment by gabeng** [comment]
Brad,
would you be willing to expose this option to the config yaml? It may be useful to other users as well.
This would apply to the GATK4 option `--max-mnp-distance` and to the `vcflib/vcfallelicprimitives` option `--use-mnps` (which is used for freebayes, if I understand correctly).

**Comment by chapmanb** [comment]
Ben;
Thanks again for following on this. I've been trying to think about the best way to incorporate MNP splitting/not-splitting in a reasonable way that we can configure and support. For your uses, it sounds like it would be most helpful if both the GATK and FreeBayes outputs did not have decomposition from the original calls at all. Is that correct? Are there other callers/approaches where you see this needed immediately as well? We've been working on incorporating Octopus (https://github.com/luntergroup/octopus) which is also aggressive on calling MNPs and phasing events.

Is just not decomposing these in the final VCF enough to make it useful for your downstream work? I recognize different callers will have different MNP representations for the same set of grouped events. Are there any harmonization tools or other approaches you're using to help mitigate this?

I appreciate you having this discussion and happy to try and figure out better ways to report and deal with these type of linked events. As a starting point, it would be incredibly useful to have truth sets where the non-decomposed events better describe the underlying biology. This would help us make decisions based on real data and tweak the approaches. Were you able to identify examples where this helped we could use as a starting point?

**Comment by gabeng** [comment]
Brad,
I must admit that I have not looked beyond GATK and FreeBayes so far and cannot comment on other variant callers. An option in bcbio to turn this on and off would enable us, though, to perform more extensive testing.
I would hesitate to turn off decomposition alltogether. As you noted, some variant callers are quite aggressively calling large MNPs. My tests so far involved settings like `--max-mnp-distance 2` to evaluate the impact on annotation with respect to codons. 
Regarding validation data: I have found some examples in GiaB manually and am currently looking into identifying them in a more automated way. I will be happy to share them with you.

**Comment by annaquaglieri16** [comment]
Hi there, I don't know if this can be useful but I have an example from other callers. I am currently trying to compare the sensitivity in recovering SNVs and INDELs from RNA-Seq using commonly used callers. As the truth set I am using a set of variants called by the km algorithm (performs local realignment and works well in RNA-Seq) + validated with sequencing (published here Lavallée 2016).

Here is an example of how Vardict, Varscan and Mutect2 from GATK3 go about finding an INDEL composed by deletion and insertions. 

![screen shot 2018-10-25 at 6 23 52 pm](https://user-images.githubusercontent.com/7087258/47483270-96acf480-d884-11e8-9cfc-bd8d2e67c673.png)

The columns `ref` , `alt` and `Location` are the ones reported by each caller (column `caller`). The columns `Mutation` and `pos` are reported by the paper and as a consequence by the `km` software https://bitbucket.org/iric-soft/km published in April. The rest is less relevant. 

As you can see, VarDict reports the MNP all together whereas Mutect2 splits into 2. Varscan reports something slightly different. 

I would love to know about any **harmonization tools** to use for comparison as @chapmanb mentioned.

Cheers, 
Anna

**Comment by gabeng** [comment]
Brad,
I can live with patching bcbio to apply the `--max-mnp-distance 2` for the time being. This is the quick fix for me. As we discussed above, the true solution would consist of annotating phased genotypes correctly. 
In my configuration the performance measured by `vcf-eval` is slightly decreased by those rare cases where two genotypes overlap. I found this to be about 5..10 occurrences per exome.
I am closing this issue now since this is probably not of general interest. Thanks for your input.
